### ISSUE TITLE
Feature: Date and DateTime classes from Ruby

### ISSUE THREAD

**Issue by AndyObtiva** [opal/opal#2628]
**Is your feature request related to a problem? Please describe.**

I cannot use `DateTime` and `Date` in my code, especially the very useful `DateTime#strptime(str, format)` method. This forces me to try to find JS alternatives, which defeats the point of using Opal Ruby.

**Describe the solution you'd like**

The full API of DateTime and Date classes from Ruby.

**Describe alternatives you've considered**

I am digging down into JS to parse Date for now.

**Additional context**

N/A

**Comment by hmdne** [comment]
Unfortunately, we won't be able to import date/datetime from Ruby, as it's implemented in C very much. What I may suggest would be to use `Date.parse` which is implemented.

Ref: #2569 which adds a bit of additional functionality

**Comment by rubyFeedback** [comment]
> What I may suggest would be to use Date.parse which is implemented.

One problem with this may be that it may not be extremely convenient to go the work-around route. Andy may
probably have to remember using Date.parse(), but even though I myself do not use DateTime#strptime, I think
it would simply be better to be able to stay "within" ruby as much as possible. One of opal's key promises is to just
write ruby and not have to worry about javascript much at all. I think the better long-term solution would be to
offer the same functionality whenever that is possible.

**Comment by AndyObtiva** [comment]
@hmdne I was thinking you'd do what the JRuby folks do. Simply re-implement `Date` and `DateTime` in Opal Ruby (or JS if you prefer) kinda like how they re-implement C features in Java or JRuby itself. There is no need to bring in the exact C code into Opal as long as you simulate it correctly in Opal Ruby (or JS if you prefer). It seems the PR that you linked is doing some of that. Ideally, you would eventually implement the complete APIs of `Date` and `DateTime`. 

That's very important to have in Opal in my opinion because Ruby Software Engineers are used to using `Date` and `DateTime` and I am sure they are used by several Ruby gems, which would become portable in Opal once `Date` and `DateTime` are fully supported.

**Comment by hmdne** [comment]
I agree with you two @rubyFeedback and @AndyObtiva . I just noted the short-term partial solutions.

**Comment by janbiedermann** [comment]
@rubyFeedback regarding #2569, when working on it and looking into ruby source to get compatibility right, i noticed, that the current approach using regexps will never be sufficient, due to various conditions/special cases. Best approach seems to port c source to js for best compatibility. Needs some time.
